diffrax/integrate.py

```diff
@@ -719,6 +719,7 @@ def diffeqsolve(
             dt0 = jnp.asarray(dt0, dtype=dtype)
         if saveat.ts is not None:
             saveat = eqx.tree_at(lambda s: s.ts, saveat, saveat.ts.astype(dtype))
+        timelikes.append(dtype)
```

Comment:
Nit: I don't think this line is necessary?

Reply:
It is not necessary for complex input, but I came across this issue with dtypes while trying to figure out the complex-type issue. Consider:
```
import jax.numpy as jnp
from jax.config import config
import diffrax
config.update("jax_enable_x64", True)

term = diffrax.ODETerm(lambda t, y, args: -1.0 * y)
y0 = jnp.array([1.0, 1.0], dtype=jnp.float32)
solver = diffrax.Euler()
sol = diffrax.diffeqsolve(term, solver, t0=0.0, t1=1.0, dt0=0.01, y0=y0)
```

https://github.com/patrick-kidger/diffrax/blob/9280c3a39c0c96851f38e4e38ab14777050f6032/diffrax/integrate.py#L593

In this case `jnp.result_type(*timelikes)` apparently will return the highest possible precision of the weak dtypes; `dtype('float64')`  if you have `jax_enable_x64`set to `True`. 

https://github.com/patrick-kidger/diffrax/blob/9280c3a39c0c96851f38e4e38ab14777050f6032/diffrax/integrate.py#L609

In this line `jnp.result_type(yi, *timelikes)` will return `dtype('float32')`  and you will have different dtypes for your timelike variables and your state and get a TypeError.

However, I see that this is a rather unrelated issue and the way I worked around it is not really ideal (if you specify your input as float32 and all other as weak types it should probably return float32 not float64). Should I open this as a new issue and suggest this workaround as a separate pull request?

Reply:
Ah -- I understand. that's a subtle issue!

I think probably the neatest fix might be to adjust [this line](https://github.com/patrick-kidger/diffrax/blob/9280c3a39c0c96851f38e4e38ab14777050f6032/diffrax/integrate.py#L609) to `jnp.result_type(yi, dtype)`?
(And either way include a comment linking to this discussion, for posterity.)

I'd be happy to take this fix alongside this PR; no need for a separate one.

On which note -- apologies for taking so long to get back around to this! If you're happy to update this to the latest version of Diffrax, then I would be happy to merge this.